### PR TITLE
Infer default sources for parameters of minimal actions

### PR DESCRIPTION
**Pull request by halter73**
This does a big chunk of #30248 while heavily refactoring the `RequestDelegateFactory` by splitting it out into a bunch of smaller methods.

For part 2 of #30248, a "scalar value type that lives in the System namespace", I went with the alternative of "Something convention-based like the presence of a bool TryParse(string, out T) method" instead. This conventions work for types in any namespace including user-defined types.

I skipped part 4 for now. Originally I was going to fall back at runtime to reading the unattributed parameter as a JSON body if there was no service of a matching type. Even though this was a runtime behavior, this didn't seem so bad at first because at least which services exist shouldn't change after app start. What I really didn't like about it in the end was that I couldn't determine if there was more than one effective "FromBody" parameter at build time. I'm still not sure how we deal with that.

@pranavkm @javiercn How does MVC inference distinguish between body and service params today?

### CODE REVIEW

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 1]
```diff
@@ -428,74 +526,170 @@ public static RequestDelegate Create(MethodInfo methodInfo, Func<HttpContext, ob
...
+                {
+                    return method;
+                }
             }
-            else if (parameter.ParameterType != valueArg.Type)
+
+            throw new Exception("static bool System.Enum.TryParse<TEnum>(string? value, out TEnum result) does not exist!!?!?");
+        }
+
+        // TODO: Use InvariantCulture where possible? Or is CurrentCulture fine because it's more flexible?
```
Comment (halter73) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:558]: I'm interested in feedback on this. One wrinkle is there isn't a standard convention for `TryParse` methods taking `IFormatProvider`. For example, `int.TryParse` and similar want a `NumberStyles` argument as well for any overload taking `IFormatProvider`. We can special-case built-in numeric types of course.

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 2]
```diff
@@ -623,14 +829,30 @@ private static class Log
...
             {
-                _requestBodyIOException(logger, exception);
+                _parameterBindingFailed(GetLogger(httpContext), parameterTypeName, parameterName, sourceValue, null);
             }
 
-            public static void RequestBodyInvalidDataException(ILogger logger, InvalidDataException exception)
+            private static ILogger GetLogger(HttpContext httpContext)
             {
-                _requestBodyInvalidDataException(logger, exception);
+                var loggerFactory = httpContext.RequestServices.GetRequiredService<ILoggerFactory>();
```
Comment (davidfowl) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:854]: Capture the logger once, this isn't free.

Comment (davidfowl) [reply src/Http/Http.Extensions/src/RequestDelegateFactory.cs:854]: Meh, not important for this change but we'll need to optimize

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 3]
```diff
@@ -428,74 +522,170 @@ public static RequestDelegate Create(MethodInfo methodInfo, Func<HttpContext, ob
...
             }
 
-            // property[key] == null ? default : (ParameterType){Type}.Parse(property[key]);
-            expr = Expression.Condition(
-                Expression.Equal(valueArg, Expression.Constant(null)),
+            Expression defaultExpression = parameter.HasDefaultValue ?
+                Expression.Constant(parameter.DefaultValue) :
+                Expression.Default(parameter.ParameterType);
+
+            // tempSourceString = httpContext.RequestValue["id];
```
Comment (davidfowl) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:669]:
```suggestion
            // tempSourceString = httpContext.RequestValue["id"];
```

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs (hunk 2 above)
Comment (halter73) [reply src/Http/Http.Extensions/src/RequestDelegateFactory.cs:854]: This optimizes the failure case quite a bit by not relying on exceptions. I think multiple parse failures are rare enough we can avoid optimizing it to that degree.

**Comment by davidfowl** [comment]
> @pranavkm @javiercn How does MVC inference distinguish between body and service params today?

MVC doesn't do this today.

File: src/Http/samples/MapActionSample/Program.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+            app.UseEndpoints(endpoints =>
+            {
+                Todo EchoTodo([FromBody] Todo todo) => todo;
+                endpoints.MapPost("/EchoTodo", (Func<Todo, Todo>)EchoTodo);
+
+                string Plaintext() => "Hello, World!";
+                endpoints.MapGet("/plaintext", (Func<string>)Plaintext);
+
+                object Json() => new { message = "Hello, World!" };
+                endpoints.MapGet("/json", (Func<object>)Json);
```
Comment (davidfowl) [inline src/Http/samples/MapActionSample/Program.cs:24]: Add one more sample showing route parameter inference:

```C#
string SayHello(string name) => $"Hello {name}";
endpoints.MapGet("/hello/{name}", (Func<string, string>)SayHello);
```

**Review by davidfowl — APPROVED** [review]
LGTM

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs (hunk 1 above)
Comment (davidfowl) [reply src/Http/Http.Extensions/src/RequestDelegateFactory.cs:553]: This is a topic that's up for debate. I think MVC does different things for different parts of the request.

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 5]
```diff
@@ -23,28 +22,39 @@ namespace Microsoft.AspNetCore.Http
...
+        private static readonly MemberExpression RequestAbortedExpr = Expression.Property(HttpContextExpr, nameof(HttpContext.RequestAborted));
+        private static readonly MemberExpression RouteValuesExpr = Expression.Property(HttpRequestExpr, nameof(HttpRequest.RouteValues));
+        private static readonly MemberExpression QueryExpr = Expression.Property(HttpRequestExpr, nameof(HttpRequest.Query));
+        private static readonly MemberExpression HeadersExpr = Expression.Property(HttpRequestExpr, nameof(HttpRequest.Headers));
+        private static readonly MemberExpression FormExpr = Expression.Property(HttpRequestExpr, nameof(HttpRequest.Form));
+        private static readonly MemberExpression StatusCodeExpr = Expression.Property(HttpResponseExpr, nameof(HttpResponse.StatusCode));
+        private static readonly MemberExpression CompletedTaskExpr = Expression.Property(null, (PropertyInfo)GetMemberInfo<Func<Task>>(() => Task.CompletedTask));
+
+        private static ConcurrentDictionary<Type, MethodInfo?> TryParseMethodCache = new();
```
Comment (campersau) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:57]:
```suggestion
        private static readonly ConcurrentDictionary<Type, MethodInfo?> TryParseMethodCache = new();
```

**Comment by halter73** [comment]
I'll submit a PR for the remaining items in https://github.com/dotnet/aspnetcore/issues/30248#issuecomment-816870890 later today. I'll add the inferred route value sample while I'm at it.

**Pull request merged**